src/AutoRest.CSharp/Common/Output/Models/Types/ObjectType.cs

```diff
@@ -33,6 +33,29 @@ protected ObjectType(string defaultNamespace, SourceInputModel? sourceInputModel
         {
         }
 
+        private bool? isUnknownDerivedType;
+        public bool IsUnknownDerivedType => isUnknownDerivedType ??= EnsureUnknownDerivedType();
+        public bool EnsureUnknownDerivedType()
+        {
+            if (!Type.Equals(Discriminator?.DefaultObjectType?.Type))
+                return false;
+
+            if (!Declaration.Name.StartsWith("Unknown", StringComparison.Ordinal))
```

Comment:
I think the challenge is in the multiple layer tree.  For example the `body-complex` project in TestServerProjects has a salmon which is a base type of SmartSalmon but the unknown type for the tree is `UnknownFish`.  Discriminator.DefaultObjectType in this case comes out to be Salmon, not UnknownFish which means that it will think this is the unknown type even though it is not.  I believe this check is fine as long as the first check above it is there as well.  We should never have a class actually named the same.